### PR TITLE
Fix grad getting enabled for parameters on which explicitly disabled in case of multiple optimizers

### LINKED ISSUE #5292: For multiple optimizers, model.toggle_optimizer() is setting the requires_grad=True for all the params.
https://github.com/PyTorchLightning/pytorch-lightning/blob/dabfeca92e0702e55f09ac53e9412672cd258cd3/pytorch_lightning/core/lightning.py#L1152-L1170

Setup : 
pytorch-lightning 1.1.2
pytorch 1.7.1

When using multiple optimizers, in the `toggle_optimizers(..)` function, the requires_grad property is set to true for all the params belonging to the param_groups of the optimizer. This is incorrect as in case the user has explicitly disabled the `requires_grad` property for some parameters permanently, the function would enable `requires_grad` for those parameters also. 

Proposed fix:
The requires_grad value for all the parameters should be stored beforehand in an object variable `self.params_dict` of lightning object. In the `toggle_optimizer()` function, instead of setting `params.requires_grad=True`, set `params.requires_grad = self.params_dict[params]`.This would set the correct value for the params of the concerned optimizer.

### PR DESCRIPTION
**Pull request by jain-anshul**
## What does this PR do?

In case of multiple optimizers, the bug causes grad to be enabled for all the parameters associated to an optimizer. This is erroneous in case there are any parameters which are explicitly kept non-trainable. 

Fixed it by storing the original grad setting for all the parameters, and then setting it accordingly when that optimizer is called.

Co-authored by : @nisheethlahoti

Fixes #5292 <- this [links related issue to this PR](https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes [if needed]?
- [ ] Did you write any new necessary tests [no need for typos, docs]? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [ ] **Check that target branch and milestone are aligned!**
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -112,6 +112,8 @@ def __init__(self, *args, **kwargs):
         self._current_hook_fx_name = None
         self._current_dataloader_idx = None
 
+        self.param_grad_dict = {}
```
Comment (tchaton) [inline pytorch_lightning/core/lightning.py:115]: Should be private.

**Review by tchaton — COMMENTED** [review]
Would you mind adding a test ?

**Comment by tchaton** [comment]
Dear @jain-anshul @nisheethlahoti,

Thanks a lot for your work ! I gave me the base to resolve the bug.

I solved in this PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/5574

Closing this one for now.

Best,
T.C
